runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowMetrics.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow;
+
+import com.google.api.services.dataflow.model.JobMetrics;
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableList;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import org.apache.beam.sdk.metrics.DistributionResult;
+import org.apache.beam.sdk.metrics.MetricFiltering;
+import org.apache.beam.sdk.metrics.MetricKey;
+import org.apache.beam.sdk.metrics.MetricName;
+import org.apache.beam.sdk.metrics.MetricQueryResults;
+import org.apache.beam.sdk.metrics.MetricResult;
+import org.apache.beam.sdk.metrics.MetricResults;
+import org.apache.beam.sdk.metrics.MetricsFilter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link MetricResults} for the Direct Runner.
+ */
+class DataflowMetrics extends MetricResults {
+  private static final Logger LOG = LoggerFactory.getLogger(DataflowMetrics.class);
+  /**
+   * Client for the Dataflow service. This can be used to query the service
+   * for information about the job.
+   */
+  private DataflowClient dataflowClient;
+
+  /**
+   * PipelineResult implementation for Dataflow Runner. It contains job state and id information.
+   */
+  private DataflowPipelineJob dataflowPipelineJob;
+
+  /**
+   * After the job has finished running, Metrics no longer will change, so their results are
+   * cached here.
+   */
+  private MetricQueryResults cachedMetricResults = null;
+
+  /**
+   * Constructor for the DataflowMetrics class.
+   * @param dataflowPipelineJob is used to get Job state and Job ID information.
+   * @param dataflowClient is used to query user metrics from the Dataflow service.
+   */
+  public DataflowMetrics(DataflowPipelineJob dataflowPipelineJob, DataflowClient dataflowClient) {
+    this.dataflowClient = dataflowClient;
+    this.dataflowPipelineJob = dataflowPipelineJob;
+  }
+
+  /**
+   * Build an immutable map that serves as a hash key for a metric update.
+   * @param metricUpdate
+   * @return
```

Comment:
Don't include Javadoc `@param` or `@return` if they aren't interesting (eg., if you have nothing to say).